### PR TITLE
update multi data interceptor and trie sync

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**

### CODE REVIEW

File: process/interceptors/multiDataInterceptor.go [hunk 1]
```diff
@@ -74,20 +75,32 @@ func (mdi *MultiDataInterceptor) ProcessReceivedMessage(message p2p.MessageP2P,
...
+
+		nodeData, ok := interceptedData.(*trie.InterceptedTrieNode)
+		if !ok {
+			mdi.throttler.EndProcessing()
+			return
 		}
 
-		err = interceptedData.CheckValidity()
+		nodeData.CreateEndOfProcessingTriggerNode()
+		_ = mdi.processor.Save(nodeData)
```
Comment (bogdan-rosianu) [inline process/interceptors/multiDataInterceptor.go]: treat the error or at least log it

File: process/interceptors/multiDataInterceptor.go [hunk 2]
```diff
@@ -74,20 +75,32 @@ func (mdi *MultiDataInterceptor) ProcessReceivedMessage(message p2p.MessageP2P,
 	lastErrEncountered := error(nil)
 	wgProcess := &sync.WaitGroup{}
 	wgProcess.Add(len(multiDataBuff))
+
+	firstDataBuff := multiDataBuff[0]
```
Comment (bogdan-rosianu) [inline process/interceptors/multiDataInterceptor.go]: you can move this to line 75

File: data/trie/interceptedNode.go [hunk 3]
```diff
@@ -79,3 +79,9 @@ func (inTn *InterceptedTrieNode) IsInterfaceNil() bool {
 func (inTn *InterceptedTrieNode) EncodedNode() []byte {
 	return inTn.encNode
 }
+
+// CreateEndOfProcessingTriggerNode changes the hash of the current node by appending the hash to the current hash.
+// This construction will be used to trigger the end of processing for all of the received data
+func (inTn *InterceptedTrieNode) CreateEndOfProcessingTriggerNode() {
+	inTn.hash = append(inTn.hash, inTn.hash...)
```
Comment (iulianpascalau) [inline data/trie/interceptedNode.go]: This is a bit messy and racy (calling append concurrently will yield unexpected results). Why not using a chan struct{}? Way safer and cleaner.

File: process/interceptors/multiDataInterceptor.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply process/interceptors/multiDataInterceptor.go]: Done.

File: process/interceptors/multiDataInterceptor.go (hunk 2 above)
Comment (BeniaminDrasovean) [reply process/interceptors/multiDataInterceptor.go]: Done.

File: data/trie/interceptedNode.go (hunk 3 above)
Comment (BeniaminDrasovean) [reply data/trie/interceptedNode.go]: The current architecture does not allow for a chan. Added a todo to use a channel at the next refactoring, and added a mutex to protect from race inconsistencies.

Comment (iulianpascalau) [reply data/trie/interceptedNode.go]: ok

File: process/interceptors/multiDataInterceptor.go [hunk 4]
```diff
@@ -71,23 +72,39 @@ func (mdi *MultiDataInterceptor) ProcessReceivedMessage(message p2p.MessageP2P,
...
+		interceptedData, err := mdi.interceptedData(firstDataBuff)
 		if err != nil {
-			lastErrEncountered = err
-			wgProcess.Done()
-			continue
+			mdi.throttler.EndProcessing()
+			return
+		}
+
+		nodeData, ok := interceptedData.(*trie.InterceptedTrieNode)
```
Comment (iulianpascalau) [inline process/interceptors/multiDataInterceptor.go]: This is used in a wrong manner.

Comment (BeniaminDrasovean) [reply process/interceptors/multiDataInterceptor.go]: Refactored.

**Pull request merged**